### PR TITLE
[Azure.Monitor.Ingestion] Add Ingestion Perf Tests

### PR DESCRIPTION
**Pull request by nisha-bhatia**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Comment by nisha-bhatia** [comment]
https://github.com/Azure/azure-sdk-for-net/issues/31758

File: sdk/monitor/Azure.Monitor.Ingestion/perf/IngestionClientTest.cs [hunk 1]
```diff
@@ -14,23 +16,42 @@ public class IngestionClientTest : PerfTest<IngestionClientTest.IngestionClientP
...
 
         public override void Run(CancellationToken cancellationToken)
         {
+            LogsIngestionClient.Upload(TestEnvironment.DCRImmutableId, TestEnvironment.StreamName, GenerateEntries(100), null, cancellationToken);
         }
 
         public override async Task RunAsync(CancellationToken cancellationToken)
         {
-            await Task.Run(() =>
+            await Task.Run(async () =>
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/IngestionClientTest.cs:52]: Why do we use Task.Run as opposed to just calling the method with await?

File: sdk/monitor/Azure.Monitor.Ingestion/perf/IngestionClientTest.cs [hunk 2]
```diff
@@ -14,23 +16,42 @@ public class IngestionClientTest : PerfTest<IngestionClientTest.IngestionClientP
     {
         /* please refer to https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/template/Azure.Template/perf/TemplateClientTest.cs to write perf test. */
 
+        protected static MonitorIngestionTestEnvironment TestEnvironment = new MonitorIngestionTestEnvironment();
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/IngestionClientTest.cs:19]: Could we possibly split this into two separate scenarios - BatchUpload (that calls convenience methods), SingleUpload that calls DPG method?

File: sdk/monitor/Azure.Monitor.Ingestion/perf/IngestionClientTest.cs [hunk 3]
```diff
@@ -14,23 +16,42 @@ public class IngestionClientTest : PerfTest<IngestionClientTest.IngestionClientP
...
+        protected readonly LogsIngestionClient LogsIngestionClient;
+        protected static List<Object> GenerateEntries(int numEntries)
+        {
+            var entries = new List<Object>();
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/IngestionClientTest.cs:23]:
```suggestion
            var entries = new List<object>();
```

**Comment by JoshLove-msft** [comment]
Should we add a perf.yml? e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf.yml

I think we'd also need a perf.bicep file e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf-resources.bicep

**Comment by nisha-bhatia** [comment]
> Should we add a perf.yml? e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf.yml
> 
> I think we'd also need a perf.bicep file e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf-resources.bicep

The bicep resource file for perf tests would be the same as the bicep file for resource creation as all the resources are used. I think then we would have duplicates of the resource bicep and tests.yml but maybe we need it for perf tests?

**Comment by JoshLove-msft** [comment]
> > Should we add a perf.yml? e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf.yml
> > I think we'd also need a perf.bicep file e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf-resources.bicep
> 
> The bicep resource file for perf tests would be the same as the bicep file for resource creation as all the resources are used. I think then we would have duplicates of the resource bicep and tests.yml but maybe we need it for perf tests?

Yes, it would probably be the same resources, but the perf pipeline is run separately from the test pipeline. We can also defer this for after GA.

File: sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            LogsIngestionClient = new LogsIngestionClient(new Uri(TestEnvironment.DCREndpoint), TestEnvironment.Credential, ConfigureClientOptions(new LogsIngestionClientOptions()));
+        }
+        public class IngestionClientPerfOptions : PerfOptions
+        {
+        }
+
+        public override void Run(CancellationToken cancellationToken)
+        {
+            LogsIngestionClient.Upload(TestEnvironment.DCRImmutableId, TestEnvironment.StreamName, data);
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs:48]: Is this the DPG method?

File: sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace Azure.Monitor.Ingestion.Perf
+{
+    public class IngestionClientTest : PerfTest<IngestionClientTest.IngestionClientPerfOptions>
+    {
+        /* please refer to https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/template/Azure.Template/perf/TemplateClientTest.cs to write perf test. */
+
+        protected static MonitorIngestionTestEnvironment TestEnvironment = new MonitorIngestionTestEnvironment();
+        protected readonly LogsIngestionClient LogsIngestionClient;
+        protected static List<object> GenerateEntries(int numEntries)
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs:21]:
```suggestion
        private static List<object> GenerateEntries(int numEntries)
```

File: sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs (hunk 4 above)
Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs:48]: yes

File: sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+using Azure.Test.Perf;
+using CommandLine;
+
+namespace Azure.Monitor.Ingestion.Perf
+{
+    public class IngestionClientTest : PerfTest<IngestionClientTest.IngestionClientPerfOptions>
+    {
+        /* please refer to https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/template/Azure.Template/perf/TemplateClientTest.cs to write perf test. */
+
+        protected static MonitorIngestionTestEnvironment TestEnvironment = new MonitorIngestionTestEnvironment();
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs:19]:
```suggestion
        private static MonitorIngestionTestEnvironment TestEnvironment = new MonitorIngestionTestEnvironment();
```

File: sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        protected readonly LogsIngestionClient LogsIngestionClient;
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs:20]:
```suggestion
        private readonly LogsIngestionClient _logsIngestionClient;
```

**Comment by nisha-bhatia** [comment]
> > > Should we add a perf.yml? e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf.yml
> > > I think we'd also need a perf.bicep file e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/perf-resources.bicep
> > 
> > 
> > The bicep resource file for perf tests would be the same as the bicep file for resource creation as all the resources are used. I think then we would have duplicates of the resource bicep and tests.yml but maybe we need it for perf tests?
> 
> Yes, it would probably be the same resources, but the perf pipeline is run separately from the test pipeline. We can also defer this for after GA.

I'll add it now!

File: sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs:48]: I thought the DPG method takes a RequestContext?

Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs:48]: public virtual Response Upload(string ruleId, string streamName, RequestContent content, RequestContext context = null) It does I am passing in null for it right now - should I make a new RequestContext and pass it in?

Comment (JoshLove-msft) [reply sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs:48]: Sorry, I meant RequestContent. Is there an implicit cast happening to RequestContent from BinaryData?

Comment (JoshLove-msft) [reply sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs:48]: Ah, TIL - https://learn.microsoft.com/en-us/dotnet/api/azure.core.requestcontent.op_implicit?view=azure-dotnet#azure-core-requestcontent-op-implicit(system-binarydata)-azure-core-requestcontent

**Comment by JoshLove-msft** [comment]
> /azp run net - monitor - perf

@mikeharder should the prepare pipelines command have created the perf pipeline?

**Comment by mikeharder** [comment]
@JoshLove-msft: No, perf pipeline creation is currently a manual process.  So far, we have been focusing on storage and messaging.  I'm not sure if we want to start adding pipelines for other scenarios yet.  Do you think this SDK is particularly perf-sensitive, or uses different codepaths than other SDKs we want to test for perf?

**Comment by JoshLove-msft** [comment]
I did find this internal doc about perf test goals - https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/417/Performance-Testing

**Comment by mikeharder** [comment]
We can certainly add a perf pipeline for this library as a one-off, if you think it's important to get these numbers (e.g. before you ship).  It will require some changes to the `azure-sdk-for-tools` repo as well.

File: sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                });
+            }
+            return entries;
+        }
+
+        public BatchUploadTest(IngestionClientPerfOptions options) : base(options)
+        {
+            LogsIngestionClient = new LogsIngestionClient(new Uri(TestEnvironment.DCREndpoint), TestEnvironment.Credential, ConfigureClientOptions(new LogsIngestionClientOptions()));
+        }
+        public class IngestionClientPerfOptions : PerfOptions
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/BatchUploadTest.cs:41]: Do we actually need these options? Can we just use the base options.

File: sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                        Computer = "Computer2",
+                        AdditionalContext = 3
+                    },
+            });
+
+        public SingleUploadTest(IngestionClientPerfOptions options) : base(options)
+        {
+            LogsIngestionClient = new LogsIngestionClient(new Uri(TestEnvironment.DCREndpoint), TestEnvironment.Credential, ConfigureClientOptions(new LogsIngestionClientOptions()));
+        }
+        public class IngestionClientPerfOptions : PerfOptions
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/perf/SingleUploadTest.cs:42]: Same comment about whether we actually need this type.

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**